lib/SIL/IR/Linker.cpp

```diff
@@ -102,7 +102,7 @@ void SILLinkerVisitor::maybeAddFunctionToWorklist(SILFunction *F,
                                                   bool setToSerializable) {
   SILLinkage linkage = F->getLinkage();
   assert((!setToSerializable || F->hasValidLinkageForFragileRef() ||
-         hasSharedVisibility(linkage)) &&
+         hasSharedVisibility(linkage) || Mod.getOptions().EmbeddedSwift) &&
```

Comment:
@eeckstein do you have guidance on whether relaxing this assert is reasonable, or if there's a better way to solve the following problem?

Basically the situation is: The stdlib has an `@_extern(c) func putchar` (without a body), which gets serialized into the stdlib. A user module *also* has a `@_extern(c) func putchar` (without a body). This then triggers this assert when we indirectly end up deserializing this forward declaration from the stdlib.

Reply:
Hm, can this also happen in regular swift?
Is it possible to narrow down the case, e.g. exclude C function decls?

Reply:
This is the minimal reproducer:
```
// RUN: %empty-directory(%t)
// RUN: %{python} %utils/split_file.py -o %t %s

// RUN: %target-swift-frontend -emit-module -o %t/MyModule.swiftmodule %t/MyModule.swift -enable-experimental-feature Embedded -parse-as-library
// RUN: %target-swift-frontend -c -I %t %t/Main.swift -enable-experimental-feature Embedded -o %t/a.o

// REQUIRES: swift_in_compiler
// REQUIRES: OS=macosx || OS=linux-gnu

// BEGIN MyModule.swift

@_extern(c)
func some_c_api()

@_transparent
public func publicFuncInAModule() {
  internalFuncInAModule()
}

@usableFromInline
internal func internalFuncInAModule() {
  some_c_api()
}

// BEGIN Main.swift

import MyModule

@_extern(c)
func some_c_api()

some_c_api()
publicFuncInAModule()
```